node/src/proxy_server/mod.rs

```diff
@@ -3911,11 +3914,71 @@ mod tests {
             record,
             &NodeRecordMetadataMessage {
                 public_key: exit_public_key,
-                unreachable_host_name_opt: Some("server.com".to_string())
+                metadata_change: NRMetadataChange::AddUnreachableHost {
+                    host_name: "server.com".to_string()
+                }
             }
         );
     }
 
+    #[test]
+    fn handle_dns_resolve_failure_does_not_sends_message_to_neighborhood_when_server_is_not_specified(
+    ) {
+        let system = System::new("test");
+        let (neighborhood_mock, _, neighborhood_log_arc) = make_recorder();
+        let cryptde = main_cryptde();
+        let mut subject = ProxyServer::new(
+            cryptde,
+            alias_cryptde(),
+            true,
+            Some(STANDARD_CONSUMING_WALLET_BALANCE),
+            false,
+        );
+        let stream_key = make_meaningless_stream_key();
+        let socket_addr = SocketAddr::from_str("1.2.3.4:5678").unwrap();
+        subject
+            .keys_and_addrs
+            .insert(stream_key.clone(), socket_addr.clone());
+        let exit_public_key = PublicKey::from(&b"exit_key"[..]);
+        let exit_wallet = make_wallet("exit wallet");
+        subject.route_ids_to_return_routes.insert(
+            1234,
+            AddReturnRouteMessage {
+                return_route_id: 1234,
+                expected_services: vec![ExpectedService::Exit(
+                    exit_public_key.clone(),
+                    exit_wallet,
+                    rate_pack(10),
+                )],
+                protocol: ProxyProtocol::HTTP,
+                server_name_opt: None,
+            },
+        );
+        let subject_addr: Addr<ProxyServer> = subject.start();
+        let dns_resolve_failure = DnsResolveFailure_0v1::new(stream_key);
+        let expired_cores_package: ExpiredCoresPackage<DnsResolveFailure_0v1> =
+            ExpiredCoresPackage::new(
+                SocketAddr::from_str("1.2.3.4:1234").unwrap(),
+                Some(make_wallet("irrelevant")),
+                return_route_with_id(cryptde, 1234),
+                dns_resolve_failure.into(),
+                0,
+            );
+        let mut peer_actors = peer_actors_builder()
+            .neighborhood(neighborhood_mock)
+            .build();
+        peer_actors.proxy_server = ProxyServer::make_subs_from(&subject_addr);
+
+        subject_addr.try_send(BindMessage { peer_actors }).unwrap();
+        subject_addr.try_send(expired_cores_package).unwrap();
+
+        System::current().stop();
+        system.run();
+        let neighborhood_recording = neighborhood_log_arc.lock().unwrap();
+        let record_opt = neighborhood_recording.get_record_opt::<NodeRecordMetadataMessage>(0);
```

Comment:
I was something that I started to write on my card GH-611. I didn't think it is used anywhere else other than in my own cards. I believe every developer has their style of writing code, so it's kinda my thing.. that I personally think.

I believe, this methodology helps us check the presence of the message, using `Some` and `None`, that I find it very useful and easy to understand. Also, we can specify the type of message this way, which is very clear.

Reply:
I see.

Although, I first started asserting on an `Option<T>` out of necessity in some of my own card. I'm delighted to see that you've also enjoyed the benefit of using it.

I find it easier to read and since it both annotates and verifies the particular message we are expecting.

In my opinion
1) It'll help the peer to understand what's going on inside the test.
2) Provide a stronger assertion.
3) It uses the feature that most of the other languages fail to provide expect Rust.

I'm still more convinced that asserting on an `Option<T>` is more appropriate. 🤝